### PR TITLE
fix: update tests for useClipboard to minimize risks of flakes

### LINKED ISSUE #13240: Test flake: Tests for `useClipboard` don't consistently flush state changes
## Issue
- Certain assertions for test cases in `useClipboard` don't consistently succeed. Previous failures:
   - https://github.com/coder/coder/actions/runs/9033841899/job/24825045193
   - https://github.com/coder/coder/actions/runs/8695401295/job/23846383155#step:4:17
- [The test file in question](https://github.com/coder/coder/blob/main/site/src/hooks/useClipboard.test.tsx)

## Notes
- Inclined to believe that the issue is with the test, rather than the functionality, because while the test is new, the functionality has mostly been unchanged since it was first created over a year ago
- Testing clipboard functionality is weird in general, so depending on what's found after digging, it might be better to make some E2E tests instead
   - Though obviously it'd be better to have a single test for the clipboard, rather than needing to have ad-hoc tests for everything that just happens to use the clipboard
   - The main issue is that React Testing Library will inject a clipboard mock the moment you try to set up any kind of user session with `user.setup`. This is great most of the time, but the problem is that the clipboard always succeeds, and there's no way to remove the mock. Once there's a single setup, the entire test is basically "tainted" with false positives
   - Because the tests actually need to verify that the fallback functionality works when a clipboard isn't available on HTTP-only connection, we can't use `user.setup` at all

### PR DESCRIPTION
**Pull request by buenos-nachos**
Closes #13240

## Changes made
- Updated all test setup/teardown logic to avoid mishaps with mock timeouts, and also to increase test isolation
- Updated test approach to use `describe.each` to improve test isolation between HTTPS and HTTP-only test cases
- Updated how mock clipboard was defined
- Added extra test cases for checking what happens when a clipboard write fully, 100% fails

### CODE REVIEW

File: site/src/hooks/useClipboard.test.tsx [hunk 1]
```diff
@@ -133,88 +135,120 @@ function renderUseClipboard(inputs: UseClipboardInput) {
...
-    }
+    console.error = (errorValue, ...rest) => {
+      const canIgnore =
+        errorValue instanceof Error &&
+        errorValue.message === COPY_FAILED_MESSAGE;
+
+      if (!canIgnore) {
+        originalConsoleError(errorValue, ...rest);
+      }
+    };
```
Comment (aslilac) [inline site/src/hooks/useClipboard.test.tsx:168]: is there a reason we can't use `spyOn` or something? feels like jest could be helping with this

**Review by aslilac — APPROVED** [review]
still pretty inscrutable to me, but maybe slightly more scrutable than before? 😵‍💫

File: site/src/hooks/useClipboard.test.tsx (hunk 1 above)
Comment (Parkreiner) [reply site/src/hooks/useClipboard.test.tsx]: Oh, yeah, good call – just wasn't thinking straight when I wrote that. Just fixed thins

**Comment by aslilac** [comment]
I think it's a pretty hard spot to scratch, and it might not be worth pouring even more time into as long as this fixes the flakes. 😅

**Pull request merged**